src/pykeen/nn/functional.py

```diff
@@ -956,6 +956,34 @@ def mure_interaction(
     ) + b_h + b_t
 
 
+def murp_interaction(
+    h: torch.FloatTensor,
+    b_h: torch.FloatTensor,
+    r_vec: torch.FloatTensor,
+    r_mat: torch.FloatTensor,
+    t: torch.FloatTensor,
+    b_t: torch.FloatTensor,
+) -> torch.FloatTensor:
+    """"""
+    h = _norm(h)
+    t = _norm(t)
+    r_mat = _norm(r_mat)
+    u_e = p_log_map(h)
+    u_W = u_e * r_vec
+    u_m = p_exp_map(u_W)
+    v_m = p_sum(t, r_mat)
+    u_m = _norm(u_m)
+    v_m = _norm(v_m)
+    sqdist = (2. * artanh(torch.clamp(torch.norm(p_sum(-u_m, v_m), 2, dim=-1), 1e-10, 1 - 1e-5))) ** 2
+
+    return -sqdist + b_h + b_t
+
+
+def _norm(h):
+    h_norm = torch.norm(h, 2, dim=-1, keepdim=True)
+    return torch.where(h_norm >= 1, h / (h_norm - 1e-5), h)
```

Comment:
Is this equivalent to `pykeen.utils.clamp_norm`?

Reply:
Hmm... I will write some code to test if they're the same since I dont want to make a mistake in reasoning about it

Reply:
Survey says: **yes**

```python
import torch

from pykeen.utils import clamp_norm


def _norm(h):
    h_norm = torch.norm(h, 2, dim=-1, keepdim=True)
    return torch.where(h_norm >= 1, h / (h_norm - 1e-5), h)


def main():
    for _ in range(20):
        x = torch.ones(5)
        torch.nn.init.normal_(x)
        print(x)

        pykeen_clamp = clamp_norm(x, maxnorm=1)
        murp_clamp = _norm(x)
        print(torch.allclose(pykeen_clamp, murp_clamp))


if __name__ == '__main__':
    main()
```